h5netcdf/core.py

```diff
@@ -234,6 +239,46 @@ def __init__(self, parent, name):
                     if k.startswith('_nc4_non_coord_'):
                         var_name = k[len('_nc4_non_coord_'):]
                     self._variables.add(var_name)
+
+        # One last pass to get the current sizes. This is necessary as h5netcdf
+        # needs to now the current size of each unlimited dimension.
+        #
+        # I'm not entirely sure how the netcdf C api determines the current
+        # size - should probably be looked into.
+        def _find_dim(h5group, dim):
+            if dim not in h5group:
+                # Might happen if the dimension has not been created yet.
+                if h5group.name == "/":
+                    return None
+                return _find_dim(h5group.parent, dim)
+            return h5group[dim]
+
+        for d in self.dimensions:
+            if self.dimensions[d] is not None:
+                continue
+            dim = _find_dim(self._h5group, d)
+            if not dim:
+                continue
+
+            # Assume the reference list at each dimension is set correctly.
+            if "REFERENCE_LIST" not in dim.attrs:
+                continue
```

Comment:
If something ended up in `self.dimensions`, it's a dimension scale *somewhere* in the group or a parent group. So we should be able to find that dimensions scale in `_find_dim()`. Possibly we need to add the check for "REFERENCE_LIST" up above into `_find_dim()`.

And if this can't happen, we shouldn't have this `continue` here. It's good to flexibly accommodate malformed input files, but we should not allow internal consistency (in h5netcdf) to fail -- that is the way of bugs.